### PR TITLE
Java chaincode implementation  for ERC721 standard token.

### PR DESCRIPTION
**Pull request by renjithpta**
Java HLF chaincode implementation of ERC721 is a standard. This standard is used for representing ownership of non-fungible tokens.

### CODE REVIEW

File: token-erc-721/chaincode-java/src/main/java/org/hyperledger/fabric/samples/erc721/models/NFT.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    return tokenURI;
+  }
+
+  /**
+   * 
+   * @param tokenURI
+   */
+  public void setTokenURI(final String tokenURI) {
+    this.tokenURI = tokenURI;
+  }
```
Comment (timo-kang) [inline token-erc-721/chaincode-java/src/main/java/org/hyperledger/fabric/samples/erc721/models/NFT.java:113]: is there any uri validation here?

File: token-erc-721/chaincode-java/src/main/java/org/hyperledger/fabric/samples/erc721/ERC721TokenContract.java [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+  /**
+   * Transfers the ownership of a non-fungible token from one owner to another owner.
+   * 
+   * @param ctx the transaction context
+   * @param from the current owner of the non-fungible token
+   * @param to the new token owner
+   * @param tokenId The non-fungible token to transfer
+   */
+  @Transaction(intent = Transaction.TYPE.SUBMIT)
+  public void transferFrom(final Context ctx, final String from, final String to,
```
Comment (timo-kang) [inline token-erc-721/chaincode-java/src/main/java/org/hyperledger/fabric/samples/erc721/ERC721TokenContract.java:167]: can ERC721 based token be transferred?

File: token-erc-721/chaincode-java/src/main/java/org/hyperledger/fabric/samples/erc721/models/NFT.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+import org.hyperledger.fabric.contract.annotation.DataType;
+import org.hyperledger.fabric.contract.annotation.Property;
+import org.json.JSONObject;
+
+import com.owlike.genson.Genson;
+import com.owlike.genson.annotation.JsonProperty;
+
+@DataType()
+public final class NFT implements Serializable {
```
Comment (timo-kang) [inline token-erc-721/chaincode-java/src/main/java/org/hyperledger/fabric/samples/erc721/models/NFT.java:16]: where can I find Nun-fungibilities on this model?

File: token-erc-721/chaincode-java/src/main/java/org/hyperledger/fabric/samples/erc721/ERC721TokenContract.java (hunk 2 above)
Comment (renjithpta) [reply token-erc-721/chaincode-java/src/main/java/org/hyperledger/fabric/samples/erc721/ERC721TokenContract.java:167]: The link for ERC721 standard is  https://eips.ethereum.org/EIPS/eip-721. 
The details of transferFrom is given below.
 /// @notice Transfer ownership of an NFT -- THE CALLER IS RESPONSIBLE
    ///  TO CONFIRM THAT `_to` IS CAPABLE OF RECEIVING NFTS OR ELSE
    ///  THEY MAY BE PERMANENTLY LOST
    /// @dev Throws unless `msg.sender` is the current owner, an authorized
    ///  operator, or the approved address for this NFT. Throws if `_from` is
    ///  not the current owner. Throws if `_to` is the zero address. Throws if
    ///  `_tokenId` is not a valid NFT.
    /// @param _from The current owner of the NFT
    /// @param _to The new owner
    /// @param _tokenId The NFT to transfer
    function transferFrom(address _from, address _to, uint256 _tokenId) external payable;

File: token-erc-721/chaincode-java/src/main/java/org/hyperledger/fabric/samples/erc721/models/NFT.java (hunk 3 above)
Comment (renjithpta) [reply token-erc-721/chaincode-java/src/main/java/org/hyperledger/fabric/samples/erc721/models/NFT.java:16]: In this sample, it is assumed that only one organization (played by Org1) is in an issuer role and can mint new tokens into their account. The mintWithTokenURI smart contract function reads the certificate information of the client identity that submitted the transaction using the GetClientIdentity.GetID() API and creates a non-fungible token associated with the client ID with the requested token ID.  The token ID  is the unique identifier that ensures non-fungibility.

**Review by jkneubuh — CHANGES_REQUESTED** [review]
Hi @renjithpta 

Thanks for the Java ERC721 contribution.  I have some minor suggestions on code formatting, style, and can not vouch for the accuracy of the routines in the PR.  But overall - it looks good.

Some really minor comments, inline below.

**Comment by jkneubuh** [comment]
One `final` comment on the review of the Java source.  If your code introduces the use of `final` modifiers, please ensure that either ALL of the immutable variables in the scope of the routines are marked as `final`, or NONE of them are declared as `final`.

Thanks @renjithpta !

**Comment by renjithpta** [comment]
I have updated the files to include the review suggestions

**Comment by renjithpta** [comment]
> One `final` comment on the review of the Java source. If your code introduces the use of `final` modifiers, please ensure that either ALL of the immutable variables in the scope of the routines are marked as `final`, or NONE of them are declared as `final`.
> 
> Thanks @renjithpta !

**Comment by renjithpta** [comment]
> One `final` comment on the review of the Java source. If your code introduces the use of `final` modifiers, please ensure that either ALL of the immutable variables in the scope of the routines are marked as `final`, or NONE of them are declared as `final`.
> 
> Thanks @renjithpta !

I have modified the code as per the review suggestion

**Review by jkneubuh — CHANGES_REQUESTED** [review]
I had a chance to run the chaincode this afternoon.  It is looking pretty good!  I have a couple of additional change requests, based on running through the erc721 README guide: 

1.  There are still a lot of empty newlines between Javadoc method comments and methods, as well as trailing newlines at the end of the method declarations. 

2.  Please change the permissions of `docker/docker-entrypoint.sh` to `0x755` to allow for execution within the Docker container. 

3.  In `Approval.java`, do not use the (deprecated) `new Boolean(boolean)` constructor, but prefer the `Boolean.valueOf(json.getBoolean("approved"))` method call (line ~115) 

4.  don't forget to add `exec` to the java targets in docker-entrypoint.sh.  This is really important for sending ^C / KILL signals to the container. 

5.  Align the contract method names with the conventions established in the contract-go and contract-typescript.  When I went through the README / guide, I noticed that the Java contract uses the typical `camelCaseConvention()`, whereas the golang and typescript examples use the `StrongCamelCaseConvention()` with a leading upcast character in the contract function names.    I found a few of these - please make a pass over all the contract function names and align them with the other samples for consistency.  e.g. : 

- `mintWithTokenURI()` --> `MintWithTokenURI()` 

- `getClientAccountBalance()` --> `ClientAccountBalance()` 

- `ownerOf()` --> `OwnerOf()` 

- `getClientAccountId()` --> `ClientAccountID()' 

- etc. ... 


This is looking great!  Apologies for the long delay getting queued up into the PR review cycle.

Thanks @renjithpta !

**Comment by jkneubuh** [comment]
Hi @renjithpta 

Something went horribly wrong with the last force push up to git.  I'm afraid this PR branch merge is no longer viable.... the last commit now shows 203 files modified in the PR.

Can you please set up a new branch on your development fork for this project, and: 

1. pull the main/latest from fabric-samples 

2. re-add / commit the chaincode-java sample source files 

3. Create a new PR / merge request. 


Let's get this project landed.  It is a really good example and demonstration of ERC-721.

**Pull request closed without merging**